packages/utilities/src/useFocusRects.ts

```diff
@@ -29,29 +29,51 @@ function setMountCounters(key: Window | HTMLElement, delta: number): number {
   return newValue;
 }
 
-function setCallbackMap(key: HTMLElement): ListenerCallbacks {
-  let callbacks = callbackMap.get(key);
+function setCallbackMap(context: IFocusRectsContext): ListenerCallbacks {
+  let callbacks = callbackMap.get(context);
   if (callbacks) {
     return callbacks;
   }
 
-  const onMouseDown = (ev: MouseEvent) => _onMouseDown(ev, key);
-  const onPointerDown = (ev: PointerEvent) => _onPointerDown(ev, key);
-  const onKeyDown = (ev: KeyboardEvent) => _onKeyDown(ev, key);
-  const onKeyUp = (ev: KeyboardEvent) => _onKeyUp(ev, key);
+  const onMouseDown = (ev: MouseEvent) => _onMouseDown(ev, context.registeredProviders);
+  const onPointerDown = (ev: PointerEvent) => _onPointerDown(ev, context.registeredProviders);
+  const onKeyDown = (ev: KeyboardEvent) => _onKeyDown(ev, context.registeredProviders);
+  const onKeyUp = (ev: KeyboardEvent) => _onKeyUp(ev, context.registeredProviders);
   callbacks = { onMouseDown, onPointerDown, onKeyDown, onKeyUp };
 
-  callbackMap.set(key, callbacks);
+  callbackMap.set(context, callbacks);
   return callbacks;
 }
 
 type AppWindow = (Window & { FabricConfig?: { disableFocusRects?: boolean } }) | undefined;
 
 export type IFocusRectsContext = {
-  providerRef?: React.RefObject<HTMLElement>;
+  /**
+   * Ref to the root element of the provider
+   */
+  readonly providerRef: React.RefObject<HTMLElement>;
+
+  /**
+   * Array of this and all child provider elements under this one in the React tree.
+   *
+   * Tracking all child providers will allow a focus event in the parent to also set focus styling in its descendants.
+   * This is needed for Combobox, for example, because the focus events happen on the parent context, but the visual
+   * focus indicator is in the combobox callout. The callout needs to be notified on focus events from the parent.
+   */
+  readonly registeredProviders: HTMLElement[];
```

Comment:
>HTMLElement[] [](http://example.com/codeflow?start=32&length=13)

Should this be a WeakMap of elements so this code doesn't potentially cause memory leaks because it's hanging on to elements that get unmounted?

Reply:
WeakMaps don't allow iterating the keys, which we need here. The provider element is added to the array when FocusRectsProvider is mounted, and removed when it's unmounted (done via `useEffect`), so it shouldn't be holding onto things after they've been unmounted.

Reply:
I was thinking about this more, and it might be possible that React would update the ref to a new HTMLElement without remounting the provider. In that case, it might be holding onto a stale HTMLElement. 

I'll update this to be an array of RefObjects instead: that will avoid holding onto the HTMLElements directly (allowing them to be garbage collected if unmounted), and make sure that we're applying the className to the ref's current value, instead of whatever was first added to the array.